src/machine/machine_stm32f405.go

```diff
@@ -32,8 +33,45 @@ type SPI struct {
 	AltFuncSelector stm32.AltFunc
 }
 
-func (spi SPI) configurePins(config SPIConfig)      {}
-func (spi SPI) getBaudRate(config SPIConfig) uint32 { return 0 }
+func (spi SPI) configurePins(config SPIConfig) {
+	config.SCK.ConfigureAltFunc(PinConfig{Mode: PinModeSPICLK}, spi.AltFuncSelector)
+	config.SDO.ConfigureAltFunc(PinConfig{Mode: PinModeSPISDO}, spi.AltFuncSelector)
+	config.SDI.ConfigureAltFunc(PinConfig{Mode: PinModeSPISDI}, spi.AltFuncSelector)
+}
+
+func (spi SPI) getBaudRate(config SPIConfig) uint32 {
+	var clock uint32
+	switch spi.Bus {
+	case stm32.SPI1:
+		clock = CPUFrequency() / 2
+	case stm32.SPI2, stm32.SPI3:
+		clock = CPUFrequency() / 4
+	}
+
+	// limit requested frequency to bus frequency
+	freq := config.Frequency
+	if freq > clock {
+		freq = clock
+	}
+
+	// calculate the exact clock divisor (freq=clock/div -> div=clock/freq).
+	// truncation is fine, since it produces a less-than-or-equal divisor, and
+	// thus a greater-than-or-equal frequency.
+	// divisors only come in consecutive powers of 2, so we can use log2 (or,
+	// equivalently, bits.Len - 1) to convert to respective enum value.
+	div := bits.Len32(clock/freq) - 1
+
+	// but DIV1 (2^0) is not permitted, as the least divisor is DIV2 (2^1), so
+	// subtract 1 from the log2 value, keeping a lower bound of 0
+	if div < 0 {
+		div = 0
+	} else if div > 0 {
+		div--
+	}
+
+	// finally, shift the enumerated value into position for SPI CR1
+	return uint32(div) << stm32.SPI_CR1_BR_Pos
+}
```

Comment:
@aykevl this is the function you previously expressed concern about, since other machine packages sometimes use a switch statement and hardcoded frequencies.

an argument supporting your concern is that this code is used by exactly one chip (STM32F405), so the fact that it is a _more generalized solution_ isn't actually benefiting anything (but in principle could be re-used elsewhere)

wanted to point this out so it isn't overlooked in review, in case the consensus would rather use the other approach.

Reply:
The value is about 6.25Mhz with the following settings.
Is this as intended?

```go
	machine.SPI0.Configure(machine.SPIConfig{
		SCK:       machine.SPI0_SCK_PIN,
		SDO:       machine.SPI0_SDO_PIN,
		SDI:       machine.SPI0_SDI_PIN,
		Frequency: 84000000,
	})
```

![ss](https://user-images.githubusercontent.com/9251039/93059738-20018600-f6ac-11ea-9e20-425f7128af01.png)

Reply:
My instrument was set up incorrectly.
At 21,000,000, it appears to be set to 41MHz.

![ss](https://user-images.githubusercontent.com/9251039/93064877-07e13500-f6b3-11ea-8cd4-c96418731548.png)

Reply:
Hmmm, the SPI0 interface is connected to hardware bus `stm32.SPI2` on `APB1`, which has a **maximum** clock freq of 42 MHz, and the smallest divisor is DIV2, so i don't think its even possible for that SPI interface to have a freq faster than 21 MHz. Can you double check this?

Reply:
My main concern here was the use of floating points and advanced math operations. While I am still not super happy about math/bits, it's not too big of a dependency so is okay.
The machine and runtime packages are a bit special when it comes to dependencies: they can't import anything that might (recursively) import the runtime. This has led to problems in the past, so I'd rather keep the list of imported packages to simple leaf packages.
(Also, right now TinyGo is rather slow and doesn't cache compiled packages. So keeping the dependency tree small also helps to reduce compile time. But this should eventually get fixed).

Regarding the 21/42 MHz: I think the 42MHz refers to the time the clock is high and the 21MHz refers to the entire pulse. So yes, the SPI peripheral is running at 21MHz (as expected).

Reply:
The following sources were used to acquire the waveforms of each.
I think stm32 is twice as fast as stm32.
Maybe the correct one is feather-m4.

feather-stm32f405 + Freq 21MHz
![image](https://user-images.githubusercontent.com/9251039/93194823-7a214a80-f783-11ea-8bf6-ffab344c7af5.png)


feather-m4 + Freq 24MHz
![image](https://user-images.githubusercontent.com/9251039/93194726-5fe76c80-f783-11ea-99ff-b771ebd461a9.png)


```go
package main

import (
	"machine"
	"time"
)

func main() {
	spi0 := machine.SPI0
	spi0.Configure(machine.SPIConfig{
		SCK:       machine.SPI0_SCK_PIN,
		SDO:       machine.SPI0_SDO_PIN,
		SDI:       machine.SPI0_SDI_PIN,
		Frequency: 21000000, // for feather-stm32f405
		//Frequency: 24000000, // for feather-m4
	})

	for {
		spi0.Tx([]byte{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}, nil)
		time.Sleep(1000 * time.Millisecond)
	}
}
```

Reply:
Can you test with an equivalent program using Arduino? That would probably be the best reference measure.

I'm thinking the Feather M4 is being driven half the frequency requested, simply because the STM32F4 datasheet indicates that SPI bus is *incapable of operating at a freq greater than 21 MHz*

Reply:
I understand.
I'm wrong and you're right.

I didn't suspect samd51, but it seems that tinygo's samd51 is wrong.

* 1 / 0.000000048 = 20.83MHz
* 1 / 0.000000084 = 11.90MHz

Arduino feather-m4 + Freq 24000000
![image](https://user-images.githubusercontent.com/9251039/93274984-9b6f4e80-f7f6-11ea-8bd4-9e9dd293b458.png)

```c

#include <SPI.h>
SPISettings MAX6675Setting (24000000, MSBFIRST, SPI_MODE0);

// the setup function runs once when you press reset or power the board
void setup() {
  // initialize digital pin LED_BUILTIN as an output.
  pinMode(LED_BUILTIN, OUTPUT);
  SPI.begin();
}

// the loop function runs over and over again forever
void loop() {
  SPI.beginTransaction(MAX6675Setting);
  SPI.transfer(0xFF);
  SPI.endTransaction();
  digitalWrite(LED_BUILTIN, HIGH);   // turn the LED on (HIGH is the voltage level)
  delay(500);                       // wait for a second
  digitalWrite(LED_BUILTIN, LOW);    // turn the LED off by making the voltage LOW
  delay(500);                       // wait for a second
}

```

Reply:
> Regarding the 21/42 MHz: I think the 42MHz refers to the time the clock is high and the 21MHz refers to the entire pulse. So yes, the SPI peripheral is running at 21MHz (as expected).

I understand.

Reply:
The SPI Frequency of atsamd will be corrected by the following PR.

#1387